bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/internal/p2/repository/helpers/RepositoryHelper.java

```diff
@@ -151,20 +150,36 @@ public static <T> IRepository<T> createMemoryComposite(IProvisioningAgent agent,
 	 */
 	@SuppressWarnings("nls")
 	public static List<Path> getSharedBundlePools() {
-		Path bundlePools = Path.of(System.getProperty("user.home"), ".p2/pools.info");
-		if (Files.isRegularFile(bundlePools)) {
-			try (Stream<String> lines = Files.lines(bundlePools, getSharedBundlePoolEncoding())) {
-				return lines.map(Path::of).filter(Files::isDirectory).toList();
+		Path agentsInfoLocation = getAgentManagerLocation().resolve("agents.info");
+		if (Files.isRegularFile(agentsInfoLocation)) {
+			Charset encoding = getSharedAgentEncoding();
+			try (Stream<String> agentLocations = Files.lines(agentsInfoLocation, encoding)) {
+				List<Path> result = new ArrayList<>();
+				agentLocations.map(agentLocation -> Path.of(agentLocation, "pools.info")).filter(Files::isRegularFile)
+						.forEach(poolsInfoLocation -> {
+							try (Stream<String> poolLocations = Files.lines(poolsInfoLocation, encoding)) {
+								poolLocations.map(poolLocation -> Path.of(poolLocation, "artifacts.xml"))
+										.filter(Files::isRegularFile).map(Path::getParent).forEach(result::add);
```

Comment:
An alternative suggestion to have no back-and-fourth on the path.
```suggestion
								poolLocations.map(Path::of)
										.filter(poolLocation -> Files.exists(poolLocation.resolve("artifacts.xml")))
										.forEach(result::add);
```

Reply:
I had the **same** thought.  What's there is slightly more concise looking, but is not really how one would express it if one wasn't using lambdas...

Reply:
I take it back.  It's not really more concise.  I will change it.

Reply:
Yes this is really a matter of taste, which may vary between persons or even in time at the same person 😅

Btw. I'm surprised that there is no constant in P2 for `artifacts.xml`?

Reply:
There 24 matches in the SDK workspace, including two in pdebuild, but only one place with private static constant that of course is only used local in that class:

org.eclipse.equinox.internal.p2.engine.ProfileMetadataRepository.ARTIFACTS_XML